zerver/lib/user_topics.py

```diff
@@ -173,26 +176,35 @@ def bulk_set_user_topic_visibility_policy_in_database(
             )
             continue
         # The request is to just 'update' the visibility policy of a topic
-        user_profiles_seeking_visibility_policy_update.append(row.user_profile)
-
-    if user_profiles_seeking_visibility_policy_update:
-        rows.filter(user_profile__in=user_profiles_seeking_visibility_policy_update).update(
-            visibility_policy=visibility_policy, last_updated=last_updated
-        )
-
-    if user_profiles_without_visibility_policy:
-        UserTopic.objects.bulk_create(
-            UserTopic(
-                user_profile=user_profile,
-                stream_id=stream_id,
-                recipient_id=recipient_id,
-                topic_name=topic_name,
-                last_updated=last_updated,
-                visibility_policy=visibility_policy,
+        user_profiles_seeking_user_topic_update_or_create.append(row.user_profile)
+
+    if user_profiles_seeking_user_topic_update_or_create:
+        query = SQL("""
+            INSERT INTO zerver_usertopic (user_profile_id, stream_id, recipient_id, topic_name, last_updated, visibility_policy)
+            SELECT * FROM UNNEST(%s) user_profile (user_profile_id)
+            CROSS JOIN (VALUES (%s, %s, %s, %s, %s)) other_values (stream_id, recipient_id, topic_name, last_updated, visibility_policy)
+            ON CONFLICT (user_profile_id, stream_id, lower(topic_name)) DO UPDATE SET
+            last_updated = EXCLUDED.last_updated,
+            visibility_policy = EXCLUDED.visibility_policy;
+        """)
```

Comment:
It might be better to use named variables rather than the ordered `%s` approach. I'm also not sure what the `other_values` name is for?

Reply:
> It might be better to use named variables rather than the ordered %s approach.

Done

> I'm also not sure what the other_values name is for?

That's just an alias for VALUES()

otherwise it results in an error:
```
django.db.utils.ProgrammingError: VALUES in FROM must have an alias
LINE 4:             CROSS JOIN (VALUES (10, 129, 'boat party', '2024...
                               ^
HINT:  For example, FROM (VALUES ...) [AS] foo.
```

Reply:
Yeah, it's a virtual table name for `VALUES`.